lib/puppet/pops/evaluator/epp_evaluator.rb

```diff
@@ -71,6 +71,20 @@ def self.evaluate(parser, func_name, scope, use_global_scope_only, parse_result,
       enforce_parameters = true
     end
 
+    # filter out all qualified names and set them in qualified_variables
+    # only pass unqualified (filtered) variable names to the the template
+    qvars = scope.compiler.qualified_variables
+    filtered_args = {}
+    template_args.each_pair do |k, v|
+      if k =~ /::/
+        k = k[2..-1] if k.start_with?('::')
+        qvars[k] = v
```

Comment:
This statment looks odd to me.
Either the `scope.compiler.qualified_variables` returns a local copy, in which case this statement is pointless since `qvars` isn't used anywhere, or this method actually modifies a hash contained by `scope.compiler.qualified_variables` in a way that will make the next call to this method treat this variable differently and put it in `filtered_args` unless it's further qualified.

Reply:
{{qualified_variables}} does not return a copy it is just a hash. There are no methods that sets/gets from there - so this is how it is done.
I don't understand your point about modifying {{qualified_variables}} and how that affects subsequent operations. The logic just sets a key/value in either a filtered hash (if key does not contain '::', or sets it in the qualified_variables otherwise (and the side effect of setting that is exactly what is wanted).
See nothing wrong here.

Reply:
So evaluating an EPP template changes global variables by poking into scope (or rather compiler) internals. Why doesn't it use`Scope#[]=` or `Scope#setvar`?